.github/workflows/run_contract_tests.yaml

```diff
@@ -17,64 +17,220 @@ jobs:
+          cd contracts/light_client 
+          cargo test -- --nocapture
+          cd ../types
           cargo test -- --nocapture
 
-      - name: Run nft_connector tests
+  test_prover:
+    needs: build_mocks
+    name: Prover test
+    runs-on: ubuntu-latest
+    steps:
+      - name: Checkout Sources
+        uses: actions/checkout@v3
+
+      - name: Download mocks
+        uses: actions/download-artifact@v3
+        with:
+          name: mocks
+      - name: Move mocks
         run: |
-          cd contracts/nft_bridge_token
-          ./build.sh
-          ./deploy.sh
-          cd ../bridge_token_deployer
-          ./build_nft.sh
-          cd ../mock_prover
-          ./build.sh
-          cd ../connector_permissions
-          ./build.sh
-          cd ../nft_connector
-          ./build.sh
-          cargo test -- --nocapture
+          mkdir contracts/mock_light_client/target
+          mkdir contracts/mock_light_client/target/wasm32-unknown-unknown
+          mkdir contracts/mock_light_client/target/wasm32-unknown-unknown/release
+          cp mock_light_client.wasm contracts/mock_light_client/target/wasm32-unknown-unknown/release
 
-      - name: Run xsc_connector tests
+      - name: Set rust version
         run: |
-          cd contracts/mock_prover
-          ./build.sh
-          cd ../connector_permissions
-          ./build.sh
-          cd ../xsc_connector
-          ./build.sh
-          cargo test -- --nocapture
\ No newline at end of file
+          rustc --version
+          rustup install 1.61
+          rustup override set 1.61
+
+      - name: Install `wasm32-unknown-unknown`
+        run: rustup target add wasm32-unknown-unknown    
+
+      - name: Run Prover Test
+        run: cd contracts/prover; ./build.sh; cargo test -- --nocapture
+
+  test_permission:
+    needs: build_mocks
+    name: Contracts Permissions test
+    runs-on: ubuntu-latest
+    steps:
+      - name: Checkout Sources
+        uses: actions/checkout@v3
+
+      - name: Download mocks
+        uses: actions/download-artifact@v3
+        with:
+          name: mocks
+      - name: Move mocks
+        run: |
+          mkdir contracts/mock_light_client/target
+          mkdir contracts/mock_light_client/target/wasm32-unknown-unknown
+          mkdir contracts/mock_light_client/target/wasm32-unknown-unknown/release
+          cp mock_light_client.wasm contracts/mock_light_client/target/wasm32-unknown-unknown/release
+          mkdir contracts/mock_prover/target
+          mkdir contracts/mock_prover/target/wasm32-unknown-unknown
+          mkdir contracts/mock_prover/target/wasm32-unknown-unknown/release
+          cp mock_prover.wasm contracts/mock_prover/target/wasm32-unknown-unknown/release
+
+      - name: Set rust version
+        run: |
+          rustc --version
+          rustup install 1.61
+          rustup override set 1.61
+
+      - name: Install `wasm32-unknown-unknown`
+        run: rustup target add wasm32-unknown-unknown    
```

Comment:
would `needs: [build_mock, test_rust] `work? 
Then copy paste rust version and wasm32 target are not needed

Reply:
it would not, in some future version we can use something like gnu-parallel to run them on the same machine to avoid rust setup multiple times:

see: https://www.gnu.org/software/parallel/